cupyx/jit/_compile.py

```diff
@@ -562,6 +562,21 @@ def _transpile_expr_internal(expr, env):
         value = _transpile_expr(expr.value, env)
         if is_constants(value):
             return Constant(getattr(value.obj, expr.attr))
+        if isinstance(value.ctype, _cuda_types.ArrayBase):
+            if 'ndim' == expr.attr:
+                return Constant(value.ctype.ndim)
+        if isinstance(value.ctype, _cuda_types.SharedMem):
+            if 'size' == expr.attr:
+                return _builtin_funcs.Len().call(value)
+        if isinstance(value.ctype, _cuda_types.CArray):
+            if 'size' == expr.attr:
+                return _builtin_funcs.Len().call(value)
+            if 'shape' == expr.attr:
+                ctype = _cuda_types.Ptr(_cuda_types.PtrDiff())
+                return Data(f'{value.code}.shape()', ctype)
+            if 'strides' == expr.attr:
+                ctype = _cuda_types.Ptr(_cuda_types.PtrDiff())
+                return Data(f'{value.code}.strides()', ctype)
```

Comment:
I'm wondering if it causes compilation error because `CArray.strides()` returns `const ptrdiff_t*` value.

Reply:
IMO, `Tuple(PtrDiff, ...)` is preferred as the return type because the shapes and the strides should not be rewritable. How about adding a new member function to `CArray` that returns a tuple value (like as `_indexing` function)?

Reply:
Does this look good to you? I can't test on my machine currently...
<details>
<summary>diff</summary>

```diff
diff --git a/cupy/_core/include/cupy/carray.cuh b/cupy/_core/include/cupy/carray.cuh
index 54d0a000b..81a1f281c 100644
--- a/cupy/_core/include/cupy/carray.cuh
+++ b/cupy/_core/include/cupy/carray.cuh
@@ -8,22 +8,7 @@
 #endif

 // Basic implementation of std::type_traits
-// We use this regardless when C++ is requested, as NVRTC by default lacks many
-// C++ features like this. We need to wrap in a namespace in case Jitify kicks
-// in and/or users provide custom definitions.
-namespace cupy {
-  namespace type_traits {
-    template<bool B, class T, class F>
-    struct conditional { typedef T type; };
-    template<class T, class F>
-    struct conditional<false, T, F> { typedef F type; };
-
-    template<bool B, class T = void>
-    struct enable_if {};
-    template<class T>
-    struct enable_if<true, T> { typedef T type; };
-  }
-}
+#include <cupy/type_traits.h>

 // math
 #ifndef M_PI
@@ -201,11 +186,42 @@ __device__ int signbit(float16 x) {return x.signbit();}
 #ifdef CUPY_JIT_MODE
 #include <cupy/swap.cuh>
 #include <cupy/tuple.cuh>
+#include <cupy/utility.h>

 template <int dim>
 struct Dim {
   __device__ Dim() {}
 };
+
+namespace helper {
+
+namespace detail {
+
+template <class T, class Seq>
+struct array_tuple_;
+template <class T, size_t... Is>
+struct array_tuple_<T, cupy::utility::index_sequence<Is...>> {
+  template <size_t>
+  using elem_type = T;
+  using type = thrust::tuple<elem_type<Is>...>;
+  static const type from(const T (&array)[sizeof...(Is)]) {
+    return thrust::make_tuple(array[Is]...);
+  }
+};
+
+} // namespace detail
+
+template <class T, size_t N>
+using array_tuple = typename detail::array_tuple_<T, cupy::utility::make_index_sequence<N>>::type;
+
+template <class T, size_t N, class R = detail::array_tuple_<
+  T, cupy::utility::make_index_sequence<N>
+>>
+typename R::type as_tuple(const T (&array)[N]) {
+  return R::from(array);
+}
+
+} // namespace helper
 #endif

 template <typename T, int _ndim, bool _c_contiguous=false, bool _use_32bit_indexing=false>
@@ -221,6 +237,10 @@ private:
   ptrdiff_t shape_[ndim];
   ptrdiff_t strides_[ndim];

+#ifdef CUPY_JIT_MODE
+  using tuple_t = helper::array_tuple<ptrdiff_t, _ndim>;
+#endif // ifdef CUPY_JIT_MODE
+
 public:
   // Constructor supports pointers or initializer lists and strides is optional
   // as long as _c_contiguous=true.
@@ -361,6 +381,14 @@ public:
   __forceinline__ __device__ T& _indexing(const Tuple &idx) {
     return const_cast<T&>(const_cast<const CArray&>(*this)._indexing(idx));
   }
+
+  __forceinline__ __device__ tuple_t _shape() const {
+    return helper::as_tuple(shape_);
+  }
+
+  __forceinline__ __device__ tuple_t _strides() const {
+    return helper::as_tuple(strides_);
+  }
 #endif

   __device__ const T& operator[](ptrdiff_t idx) const {
diff --git a/cupy/_core/include/cupy/type_traits.h b/cupy/_core/include/cupy/type_traits.h
new file mode 100644
index 000000000..90adcca6f
--- /dev/null
+++ b/cupy/_core/include/cupy/type_traits.h
@@ -0,0 +1,29 @@
+/*
+ * Basic implementation of std::type_traits
+ * We use this regardless when C++ is requested, as NVRTC by default lacks many
+ * C++ features like this. We need to wrap in a namespace in case Jitify kicks
+ * in and/or users provide custom definitions.
+ */
+
+namespace cupy {
+  namespace type_traits {
+    template<bool B, class T, class F>
+    struct conditional { typedef T type; };
+    template<class T, class F>
+    struct conditional<false, T, F> { typedef F type; };
+
+    template<bool B, class T = void>
+    struct enable_if {};
+    template<class T>
+    struct enable_if<true, T> { typedef T type; };
+
+    template <class T, T v>
+    struct integral_constant {
+      static const T value = v;
+      using value_type = T;
+      using type = integral_constant;
+      const operator value_type() const noexcept { return value; }
+      const value_type operator()() const noexcept { return value; }
+    };
+  }
+}
diff --git a/cupy/_core/include/cupy/utility.h b/cupy/_core/include/cupy/utility.h
new file mode 100644
index 000000000..d450ffb23
--- /dev/null
+++ b/cupy/_core/include/cupy/utility.h
@@ -0,0 +1,69 @@
+#include <cupy/type_traits.h>
+
+namespace cupy {
+
+namespace utility {
+
+template <class T, T... Ints>
+struct integer_sequence {
+  using value_type = T;
+  static const size_t size() noexcept { return sizeof...(Ints); }
+};
+
+template <size_t... Ints>
+using index_sequence = integer_sequence<size_t, Ints...>;
+
+namespace detail {
+
+template <class Offset, class Lhs, class Rhs>
+struct merge_sequence_ {};
+template <class T, T Offset, T... Lhs, T... Rhs>
+struct merge_sequence_<
+  type_traits::integral_constant<T, Offset>,
+  integer_sequence<T, Lhs...>, integer_sequence<T, Rhs...>
+> {
+  using type = integer_sequence<T, Lhs..., (Offset + Rhs)...>;
+};
+
+template <class T, class N>
+struct make_integer_sequence_ {
+  using L = type_traits::integral_constant<T, N::value / 2>;
+  using R = type_traits::integral_constant<T, N::value - L::value>;
+  using type = typename merge_sequence_<
+    L, typename make_integer_sequence_<T, L>::type, typename make_integer_sequence_<T, R>::type
+  >::type;
+};
+template <class T>
+struct make_integer_sequence_<T, type_traits::integral_constant<T, 0>> {
+  using type = integer_sequence<T>;
+};
+template <class T>
+struct make_integer_sequence_<T, type_traits::integral_constant<T, 1>> {
+  using type = integer_sequence<T, 0>;
+};
+template <class T>
+struct make_integer_sequence_<T, type_traits::integral_constant<T, 2>> {
+  using type = integer_sequence<T, 0, 1>;
+};
+template <class T>
+struct make_integer_sequence_<T, type_traits::integral_constant<T, 3>> {
+  using type = integer_sequence<T, 0, 1, 2>;
+};
+template <class T>
+struct make_integer_sequence_<T, type_traits::integral_constant<T, 4>> {
+  using type = integer_sequence<T, 0, 1, 2, 3>;
+};
+
+} // namespace detail
+
+template <class T, T N>
+using make_integer_sequence = typename detail::make_integer_sequence_<
+  T, type_traits::integral_constant<T, N>
+>::type;
+
+template <size_t N>
+using make_index_sequence = make_integer_sequence<size_t, N>;
+
+} // namespace utility
+
+} // namespace cupy
```
</details>

Reply:
An implementation like as follows may be simpler. (This is a pseudo-code, and I have not tried compilation...)
```cpp
template <int ndim, typename... Args>
struct as_tuple {
    using ChildClass = as_tuple<ndim - 1, ptrdiff_t, Args...>;
    using T = typename ChildClass::T;

    template <typename Ints>
    static T call(Ints ints, Args... args) {
        return ChildClass::call(ints, ints[ndim - 1], args);
    }
};

template <typename... Args>
struct as_tuple<0, Args...> {
    using T = thrust::tuple<Args...>;

    template <typename Ints>
    static T call(Ints ints, Args... args) {
        return thrust::make_tuple(args...);
    }
};

...

return as_tuple<ndim>::call(shape)
```

Reply:
I just realized that though tuple ensures constancy, it also prohibits non-const indexing. So maybe the simplest way is just add const qualifier to the pointer.

Reply:
Do you mean that CuPy JIT will handle the shape as of an array type? Or internally holds the shape values as an array, and the JIT compiler handles it as a tuple of ints?

In any case, I don't want to directly feed back the compilation error from the generated CUDA code to the users for usability. CuPy's JIT compiler should detect the compilation error before feeding a generated code to the CUDA compiler.

Reply:
In addition, a shape of tuple of ints enables us to use pattern match. As a typical use case, we can write like `n, m = array.shape()`. :smiley:

Reply:
If JIT handle the shape as tuple, then we can't write `for i in range(array.ndim): array.shape[i]`. Maybe returning a 1-dim `CArray` can cover both usecases.

Reply:
> So you are planning to support a fixed-length array type whose length is determined at compile time?

Yes, just like `std::array`.
> It seems that we can cover these use cases by adding `size` and `writable` attributes to `Ptr` type.

That's a great idea. Thanks~

Reply:
I added the following lines in `_transpile_stmt` before calling `_transpile_lvalue`:
https://github.com/cupy/cupy/blob/f0c999e8d22880f06e7f9bcf260209e702688be9/cupyx/jit/_compile.py#L348-L355
Maybe there's a better way to support unpacking `Ptr` without calling `thrust::make_tuple`?

Reply:
There are several concerns:
- The same expression `value.code` seems to be evaluated multiple times.
- Nested tuples are hard to support (e.g., `(n1, n2), (s1, s2) = array.shape, array.stride`)

Now I'm again feeling that it would be better to return a tuple as the shape. Since `array.ndim` is a constant, we can write `for i in range(array.ndim): array.shape[i]` by supporting compile-time loop unrolling in the future.

Anyway, the support for `.shape` and `.strides` still needs discussion will take some time. How about starting from the support for `len`, `min`, `max` and `.ndim` in this PR (It seems to be able to merge them soon! :smiley:), and continuing that discussion in another issue or PR?

Reply:
I'll close this one and reopen another pr for `len`, `min`, `max` and `ndim`.

Reply:
Anyway currently we don't support unpacking nested tuples. I'm thinking that instead of using `thrust::tie` on the lhs, we should call `thrust::get` on the rhs and treat each pair of elements as an assignment. Therefor `(a, b), (c, d) = (1, 2), (3, 4)` should be transpiled as `a, b = 1, 2` & `c, d = 3, 4`, and then `a = 1`, `b = 2`, `c = 3` & `d = 4`. To avoid calling rhs multiple times (if the rhs is an rvalue), maybe we should first assign it to a temporary variable and enclose the entire assignments in a block scope. If I recall correctly we always declare the variables at the beginning of the function, so the additional block scope won't introduce any side effect.

Reply:
> Anyway currently we don't support unpacking nested tuples.

Thank you for the reporting! I didn't notice that. I will fix it soon!